src/libsyntax/ast.rs

```diff
@@ -166,30 +166,14 @@ impl GenericArgs {
     }
 }
 
-#[derive(Clone, RustcEncodable, RustcDecodable, Debug)]
-pub enum GenericArg {
-    Lifetime(Lifetime),
-    Type(P<Ty>),
-    Const(AnonConst),
-}
-
-impl GenericArg {
-    pub fn span(&self) -> Span {
-        match self {
-            GenericArg::Lifetime(lt) => lt.ident.span,
-            GenericArg::Type(ty) => ty.span,
-            GenericArg::Const(ct) => ct.value.span,
-        }
-    }
-}
-
 /// A path like `Foo<'a, T>`.
 #[derive(Clone, RustcEncodable, RustcDecodable, Debug, Default)]
 pub struct AngleBracketedArgs {
     /// The overall span.
     pub span: Span,
-    /// The arguments for this path segment.
-    pub args: Vec<GenericArg>,
+    pub lifetimes: Vec<Lifetime>,
+    pub types: Vec<P<Ty>>,
+    pub const_args: Vec<AnonConst>,
```

Comment:
How are you going to preserve the correct order? Types and constants can be interleaved, last I checked.

Reply:
(They're currently blocked from being so, but this restriction was not necessarily intended to be permanent.)